### PR TITLE
Bambda to Filter Authenticated Requests

### PR DESCRIPTION
**Pull request by joe-ds**
### Bambda Contributions

* [x] Bambda has a valid [header](https://github.com/PortSwigger/bambdas/blob/73077e7ff3f6fac9db7dc95c0a00bd842b6bb64c/Proxy/HTTP/FilterOnCookieValue.bambda#L1-L5), featuring an `@author` annotation and suitable description
* [x] Bambda compiles and executes as expected
* [x] Only .bambda files have been added or modified (README.md files are automatically updated / generated after PR merge)

This is a Bambda to filter 200 OK authenticated requests. It has four variables near the top. By default it looks for Authorization headers, but setting a session cookie name (and optionally value) means it also checks for these.

### CODE REVIEW

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+var sessionCookieValue = "";      // If given, will check if cookie with sessionCookieName has this value.
+
+var request = requestResponse.request();
+var response = requestResponse.response();
+var mimeType = requestResponse.mimeType();
+var path = requestResponse.request().pathWithoutQuery().toLowerCase();
+
+var inScope = requestResponse.request().isInScope();
+
+var isAuthorised = response.isStatusCodeClass(StatusCodeClass.CLASS_2XX_SUCCESS);
```
Comment (portswigger-penguin) [inline Proxy/HTTP/FilterAuthenticated.bambda:23]: Potentially could do an early return and take the `isAuthorised` check out of the final condition:
```java
// return false if response is not authorized
if (!response.isStatusCodeClass(StatusCodeClass.CLASS_2XX_SUCCESS)) {
	return false;
}
```

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+var configNotInScopeOnly = true;  // If set to false, won't show out-of-scope items.
+var sessionCookieName = "";       // If given, will look for a cookie with that name.
+var sessionCookieValue = "";      // If given, will check if cookie with sessionCookieName has this value.
+
+var request = requestResponse.request();
+var response = requestResponse.response();
+var mimeType = requestResponse.mimeType();
+var path = requestResponse.request().pathWithoutQuery().toLowerCase();
+
+var inScope = requestResponse.request().isInScope();
```
Comment (portswigger-penguin) [inline Proxy/HTTP/FilterAuthenticated.bambda:21]: You can reuse the `request` variable here.

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+var configNoFilter = true;        // If set to false, won't show JS, GIF, JPG, PNG, CSS.
+var configNotInScopeOnly = true;  // If set to false, won't show out-of-scope items.
+var sessionCookieName = "";       // If given, will look for a cookie with that name.
+var sessionCookieValue = "";      // If given, will check if cookie with sessionCookieName has this value.
+
+var request = requestResponse.request();
+var response = requestResponse.response();
+var mimeType = requestResponse.mimeType();
+var path = requestResponse.request().pathWithoutQuery().toLowerCase();
```
Comment (portswigger-penguin) [inline Proxy/HTTP/FilterAuthenticated.bambda:19]: These 2 variables can move to be closer to where they are used to calculate the filterDenyList.

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	        sessionCookie = true;
+        } else {
+            sessionCookie = false;
+        }
+    } else {
+        sessionCookie = false;
+    };
+} else {
+    sessionCookie = false;
+}
```
[the remark this answers is not in the document]
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: That's much better! I've made the change, and tested it and it works.

File: Proxy/HTTP/FilterAuthenticated.bambda (hunk 2 above)
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: Done.

File: Proxy/HTTP/FilterAuthenticated.bambda (hunk 3 above)
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: Moved down to lines 27 and 28.

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 5]
```diff
@@ -0,0 +1,48 @@
+/**
+ * Filters authenticated 200 OK requests in Proxy HTTP history. See four config values below.
+ *
+ * @author joe-ds (https://github.com/joe-ds)
+ **/
+
+var request = requestResponse.request();
+var response = requestResponse.response();
+
+if (!response.isStatusCodeClass(StatusCodeClass.CLASS_2XX_SUCCESS) || !requestResponse.hasResponse()) {
```
[the remark this answers is not in the document]
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: Oops! Good catch, thank you. I've split the check into two if statements.

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+var inScope = request.isInScope();
+var authHeader = request.hasHeader("Authorization");
+
+boolean sessionCookie = request.headerValue("Cookie") != null
+                            && !sessionCookieName.isEmpty()
+                            && request.hasParameter(sessionCookieName, HttpParameterType.COOKIE) 
+			                && (sessionCookieValue.isEmpty() || sessionCookieValue.equals(request.parameter(sessionCookieName, HttpParameterType.COOKIE).value()));
+
+var path = requestResponse.request().pathWithoutQuery().toLowerCase();
```
[the remark this answers is not in the document]
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: Done.

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+if (!response.isStatusCodeClass(StatusCodeClass.CLASS_2XX_SUCCESS) || !requestResponse.hasResponse()) {
+    return false;
+}
+
+var configNoFilter = true;        // If set to false, won't show JS, GIF, JPG, PNG, CSS.
+var configNotInScopeOnly = true;  // If set to false, won't show out-of-scope items.
+var sessionCookieName = "";       // If given, will look for a cookie with that name.
+var sessionCookieValue = "";      // If given, will check if cookie with sessionCookieName has this value.
+
+var inScope = request.isInScope();
```
[the remark this answers is not in the document]
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: This is now inlined.

File: Proxy/HTTP/FilterAuthenticated.bambda [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+var response = requestResponse.response();
+
+if (!response.isStatusCodeClass(StatusCodeClass.CLASS_2XX_SUCCESS)) {
+    return false;
+}
+
+var configNoFilter = true;        // If set to false, won't show JS, GIF, JPG, PNG, CSS.
+var configNotInScopeOnly = true;  // If set to false, won't show out-of-scope items.
+var sessionCookieName = "";       // If given, will look for a cookie with that name.
+var sessionCookieValue = "";      // If given, will check if cookie with sessionCookieName has this value.
```
Comment (portswigger-penguin) [inline Proxy/HTTP/FilterAuthenticated.bambda:21]: One more thing. In general we have a pattern of putting config variables at the top of the bambda just for easier editing for other users.

**Review by portswigger-penguin — CHANGES_REQUESTED** [review]
Looks good. Happy to approve once the config variables are at the top of the file.

File: Proxy/HTTP/FilterAuthenticated.bambda (hunk 8 above)
Comment (joe-ds) [reply Proxy/HTTP/FilterAuthenticated.bambda]: Ah, that makes sense. I've moved them now.

Thanks for taking the time to make these change requests so I get it right!

**Review by portswigger-penguin — APPROVED** [review]

**Review by Hannah-PortSwigger — APPROVED** [review]
Thanks for your efforts! Looks good 👍

**Comment by joe-ds** [comment]
Thank you for your  very helpful corrections and patience!

**Pull request merged**